design/src/rfcs/rfc0022_error_context_and_compatibility.md

```diff
@@ -0,0 +1,370 @@
+RFC: Error Context and Compatibility
+====================================
+
+> Status: RFC
+>
+> Applies to: Generated clients and shared rust-runtime crates
+
+This RFC proposes a pattern for writing Rust errors to provide consistent
+error context AND forwards/backwards compatibility. The goal is to strike
+a balance between these four goals:
+
+1. Errors are forwards compatible, and changes to errors are backwards compatible
+2. Errors are ergonomic
+3. Error messages are easy to debug
+4. Errors adhere to Rust's `Error` trait
+
+_Note:_ This RFC is _not_ about error backwards compatibility when it comes to error serialization/deserialization
+for transfer over the wire. The Smithy protocols cover that aspect.
+
+Past approaches in smithy-rs
+----------------------------
+
+This section examines some examples found in `aws-config` that illustrate different problems
+that this RFC will attempt to solve, and calls out what was done well, and what could be improved upon.
+
+### Case study: `InvalidFullUriError`
+
+To start, let's examine `InvalidFullUriError` (doc comments omitted):
+```rust
+#[derive(Debug)]
+#[non_exhaustive]
+pub enum InvalidFullUriError {
+    #[non_exhaustive] InvalidUri(InvalidUri),
+    #[non_exhaustive] NoDnsService,
+    #[non_exhaustive] MissingHost,
+    #[non_exhaustive] NotLoopback,
+    DnsLookupFailed(io::Error),
+}
+
+impl Display for InvalidFullUriError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            InvalidFullUriError::InvalidUri(err) => write!(f, "URI was invalid: {}", err),
+            InvalidFullUriError::MissingHost => write!(f, "URI did not specify a host"),
+            // ... omitted ...
+        }
+    }
+}
+
+impl Error for InvalidFullUriError {
+    fn source(&self) -> Option<&(dyn Error + 'static)> {
+        match self {
+            InvalidFullUriError::InvalidUri(err) => Some(err),
+            InvalidFullUriError::DnsLookupFailed(err) => Some(err),
+            _ => None,
+        }
+    }
+}
+```
+
+This error does a few things well:
+1. Using `#[non_exhaustive]` on the enum allows new errors to be added in the future.
+2. Breaking out different error types allows for more useful error messages, potentially with error-specific context.
+   Customers can match on these different error variants to change their program flow, although it's not immediately
+   obvious if such use cases exist for this error.
+4. The error cause is available through the `Error::source()` impl for variants that have a cause.
+
+However, there are also a number of things that could be improved:
+1. All tuple/struct enum members are public, and `InvalidUri` is an error from the `http` crate.
+   Exposing a type from another crate can potentially lock the GA SDK into a specific crate version
+   if breaking changes are ever made to the exposed types.
```

Comment:
Not to mention, this makes it more annoying to use something like Surf which uses a different HTTP library